cli/cli.go

```diff
@@ -1,456 +1,822 @@
+			log.Warnf("unsupported install dependencies %s", strings.Join(unsupported, " "))
+			return nil
+		}
+		if err := confirmInstallDependencies(p, o); err != nil {
+			return err
+		}
+		if !o.installDependencies {
+			for _, s := range supported {
+				if err := utils.ShowInstructions(a.GetCommandRunner(), s); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+		for _, s := range supported {
+			if err := utils.InstallDependency(a.GetCommandRunner(), s); err != nil {
+				return err
+			}
+		}
+	}
+	var services []string
+	switch o.nodeType {
+	case NodeTypeFullNode:
+		services = []string{"execution", "consensus"}
+		if o.withValidator {
+			services = append(services, "validator")
+		}
+	case NodeTypeExecution:
+		services = []string{"execution"}
+	case NodeTypeConsensus:
+		services = []string{"consensus"}
+	case NodeTypeValidator:
+		services = []string{"validator"}
+	}
+	run, err := p.Confirm("Run services now?", false)
 	if err != nil {
-		return nil, fmt.Errorf(configs.InvalidGenerationPath, flags.generationPath, err)
+		return err
+	}
+	if run {
+		if err := a.SetupContainers(actions.SetupContainersOptions{
+			GenerationPath: o.generationPath,
+			Services:       services,
+		}); err != nil {
+			return err
+		}
+		if err := a.RunContainers(actions.RunContainersOptions{
+			GenerationPath: o.generationPath,
+			Services:       services,
+		}); err != nil {
+			return err
+		}
+		// TODO: Final tips
+		log.Error("show final tips")
 	}
-
-	return &clientImages, nil
+	return nil
 }
 
-func runCliCmd(cmd *cobra.Command, args []string, flags *CliCmdFlags, clientImages *clientImages, cmdRunner commands.CommandRunner, prompt prompts.Prompt, serviceManager services.ServiceManager, sedgeActions actions.SedgeActions) []error {
-	// Warnings
-	// Warn if custom images are used
-	if clientImages.execution != "" {
-		log.Warn(configs.CustomExecutionImagesWarning)
+func generateKeystore(p prompter.Prompter, o *CliCmdOptions, a actions.SedgeActions) error {
+	if err := selectKeystoreSource(p, o); err != nil {
+		return err
 	}
-	if clientImages.consensus != "" {
-		log.Warn(configs.CustomConsensusImagesWarning)
+	switch o.keystoreSourceType {
+	case SourceTypeSkip:
+		return nil
+	case SourceTypeCreate:
+		// Get the mnemonic
+		if err := selectKeystoreMnemonicSource(p, o); err != nil {
+			return err
+		}
+		switch o.keystoreMnemonicSource {
+		case SourceTypeCreate:
+			candidate, err := keystores.CreateMnemonic()
+			if err != nil {
+				return err
+			}
+			if err := saveMnemonic(a.GetCommandRunner(), candidate); err != nil {
+				return err
+			}
+			o.keystoreMnemonic = candidate
+		case SourceTypeExisting:
+			if err := inputKeystoreMnemonicPath(p, o); err != nil {
+				return err
+			}
+		}
+		// Get the passphrase
+		if err := selectKeystorePassphraseSource(p, o); err != nil {
+			return err
+		}
+		switch o.keystorePassphraseSource {
+		case SourceTypeExisting:
+			if err := inputKeystorePassphrasePath(p, o); err != nil {
```

Comment:
Same or similar here. Give the user the chance to amend any error. It is painful to get this far in the setup and re-do everything even when the files have already been generated.